editor.planx.uk/src/ui/shared/CharacterCounter.tsx

```diff
@@ -0,0 +1,71 @@
+import Typography from "@mui/material/Typography";
+import { visuallyHidden } from "@mui/utils";
+import { debounce } from "lodash";
+import React, { useCallback, useEffect, useState } from "react";
+import { FONT_WEIGHT_SEMI_BOLD } from "theme";
+
+export type Props = {
+  limit: number;
+  count: number;
+};
+
+export const CharacterCounter: React.FC<Props> = ({ limit, count }) => {
+  const [screenReaderCount, setScreenReaderCount] = useState<number>(0);
+  const [showReaderCount, setShowReaderCount] = useState<boolean>(false);
+
+  const updateCharacterCount = useCallback(
+    debounce((count: number) => {
+      setScreenReaderCount(count);
+      setShowReaderCount(true);
+    }, 500),
+    []
+  );
+
+  const currentCharacterCount = limit - count;
+  const showCharacterLimitError = currentCharacterCount < 0;
+
+  useEffect(() => {
+    if (count !== screenReaderCount) {
+      setShowReaderCount(false);
+    }
+    updateCharacterCount(currentCharacterCount);
+  }, [currentCharacterCount, updateCharacterCount, count, screenReaderCount]);
+
+  const characterLimitText = showCharacterLimitError
+    ? `You have ${Math.abs(currentCharacterCount)} characters too many`
+    : `You have ${currentCharacterCount} characters remaining`;
+
+  const screenReaderCountText = showCharacterLimitError
+    ? `You have ${Math.abs(screenReaderCount)} characters too many`
+    : `You have ${screenReaderCount} characters remaining`;
+
+  return (
+    <>
+      <Typography
+        id={"character-hint"}
+        sx={visuallyHidden}
+        aria-hidden={"true"}
+      >
+        {`You can enter up to ${limit} characters`}
+      </Typography>
+      <Typography
+        paddingTop={0.5}
+        variant="body2"
+        color={showCharacterLimitError ? "error" : "InfoText"}
+        fontWeight={showCharacterLimitError ? FONT_WEIGHT_SEMI_BOLD : 400}
+        id={"character-live-hint"}
+        aria-hidden={"true"}
+      >
+        {characterLimitText}
+      </Typography>
+      <Typography
+        aria-live="polite"
+        aria-atomic="true"
+        sx={visuallyHidden}
+        data-testid="screen-reader-count"
+      >
+        {showReaderCount && screenReaderCountText}
```

Comment:
This is hiding and redisplaying the screen reader announcement, it works to ensure the screen reader only reads relevant info to the user, but it leads to the Label being re-announced multiple times.

A way around this would be to separate out a label as I have done for the `"character-hint"` while hiding the current label. This would allow re-rendering of the announcement without the issue of extra noise.

Interesting to know if I am over-thinking this and adding too many layers of complexity as the who point of the label for an input is to ease this type of thing, but I couldn't think of anything else.

Reply:
I'd expect that because of the debounce, we don't need the variable `showReaderCount` - it's just a single value which is updated in place. Can we remove this guard and just keep `screenReaderCountText` in place?

Reply:
Tried to remove this now `<CharacterCounter>` sits outwith the `<Input>` but still found it announcing it too many times when typing.